Spring Security Bonus Module (#54)
LaunchCodeEducation/java-web-dev-curriculum: content/authentication/next-steps/bonus-module/spring-security/_index.md

**TODO**

## Implementing Spring Security 6 for Authentication - TEXT

To begin, we need to make sure that our Gradle project includes the necessary
dependencies for Spring Security 6.

Update the `build.gradle` file to include the following packages in the
`dependencies` object. You will also need to **remove** the
`spring-security-crypto:5.5.1` package

```groovy
    // Remove implementation("org.springframework.security:spring-security-crypto:5.5.1")
	implementation 'org.thymeleaf.extras:thymeleaf-extras-springsecurity6'
	implementation 'org.springframework.boot:spring-boot-starter-security'
```

This lesson will consist of four parts of implementation to adopt the
Spring Security framework:

1. Refactor `UserService` to implement Spring Security `UserDetailsService`
interface methods that take care of authentication and user session
1. Remove `AuthenticationFilter` and `WebApplicationConfig` to replace with
Spring Security 6 `WebSecurity` class, which handles authorization and
request filtering
1. Refactor other controllers and services to make use of updated `UserService`
1. Improve error pages and error redirection

### Authentication in Spring Security 6

> First, it is important to take a look at the [Spring Security Authentication

Reviewer: ```suggestion
Before diving in, take a look at the [Spring Security Authentication
```

classDiagram
    UserDetailsService <|-- IUserService : extends
    IUserService <|.. UserService : implements
    class UserDetailsService{
        loadUserByUsername(String username) UserDetails
    }
    class IUserService{
        findByUsername(String username) User
        save(RegisterFormDTO registration) User
        getCurrentUser() User
        findById(Integer id) Optional~User~
        findAll() List~User~
        deleteUser(Integer id) User
    }
    class UserService{
        loadUserByUsername(String username) UserDetails
        save(RegisterFormDTO registration) User
        getCurrentUser() User
        findByUsername(String username) User
        findById(Integer id) Optional~User~
        findAll() List~User~
        deleteUser(Integer id) User
    }
```

Create a new *interface* in the `codingevents.services` package named
`IUserService`.

This interface should **extend** the `UserDetailsService` interface, which
is found in the `org.springframework.security.core.userdetails` package.

`IUserService` should include two method definitions, which mirror the
methods we have already implemented in `UserService`.

* `User findByUsername(String username);`

> * `User save(RegisterFormDTO registration);`

Reviewer: I would encourage you to turn some of the unordered lists into ordered lists. Ordered lists are easier for screenreader users to navigate.

To begin, we can **remove** the `AuthenticationFilter` class entirely, as well
as the `WebApplicationConfig` class.

Next, create a new class `WebSecurity` in the `security` package. This class
will set up the `SecurityFilterChain` described in the [Authorizing
Requests](https://docs.spring.io/spring-security/reference/servlet/authorization/authorize-http-requests.html#authorize-requests)
section of the documentation.

This class should also include `UserService` and `PasswordEncoder` autowired
fields.

```java
@Configuration
@EnableWebSecurity
@EnableMethodSecurity
@EnableTransactionManagement
public class WebSecurity {

    @Autowired
    private UserService userService;

    @Autowired
    private PasswordEncoder passwordEncoder;
}
```

1. `@Configuration` annotation lets Spring know that this class will set up
a `@Bean`
1. `@EnableWebSecurity` annotation lets Spring know that this class will set
up the `SecurityFilterChain` bean
1. `@EnableMethodSecurity` annotation gives the ability to use some helpful
annotations to secure specific controller methods
1. `@EnableTransactionManagement` annotation allows our Spring project to use
the `@Transactional` annotation elsewhere

> To set up the `SecurityFilterChain` we will use method chaining to build an

Reviewer: ```suggestion
To set up the `SecurityFilterChain`, we will use method chaining to build an
```

            errors.rejectValue("username", "bad.credentials", "Invalid e-mail or password");
            return "/login";
        }
```

Last, weirdly enough, we can remove our request handler for logout, as we
included the `/logout` route as part of our security filter setup in
`WebSecurity`. Navigating to `/logout` will invalidate the security session
automatically.

#### Update `EventCategoryController`

To update `EventCategoryController`, we want to remove the reference to
`authController` that is used to retrieve the current user. Instead we'll
make a call to a `eventCategoryService` method that assumes the current
user.

In `EventCategoryController`, remove the field that references
`AuthenticationController`, and update the `displayAllCategories` method to
call `eventCategoryService.getAllCategoriesByCurrentUser()`, like below:

```java
    @GetMapping
    public String displayAllCategories(Model model, HttpSession session) {
        model.addAttribute("title", "All Categories");
        model.addAttribute("categories", eventCategoryService.getAllCategoriesByCurrentUser());
        return "eventCategories/index";
    }
```

#### Update `EventController`

> Our goal for `EventController` will be the same as `EventCategoryController`
> --- we want to remove all references to `AuthenticationController` and the

Reviewer: ```suggestion
Our goal for `EventController` will be the same as `EventCategoryController`: to remove all references to `AuthenticationController` and the
```

code that was necessary to grab the current user. Instead, we will make
calls to `eventService` and `eventCategoryService` methods that assume the
current user.

First, **remove** the `AuthenticationController` autowired field.

> Next, we need one new method in `EventCategoryService` for retrieving a
> category by an id but also assuming the current user. We want to prevent

Reviewer: ```suggestion
Next, we need one new method in `EventCategoryService` for retrieving a
category that the current user created by the category id. We want to prevent
```
